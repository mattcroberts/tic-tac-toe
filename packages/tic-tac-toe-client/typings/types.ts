/* tslint:disable */
import { GraphQLResolveInfo } from "graphql";

export type Resolver<Result, Parent = any, Context = any, Args = any> = (
    parent: Parent,
    args: Args,
    context: Context,
    info: GraphQLResolveInfo
) => Promise<Result> | Result;

export type SubscriptionResolver<
    Result,
    Parent = any,
    Context = any,
    Args = any
> = {
    subscribe<R = Result, P = Parent>(
        parent: P,
        args: Args,
        context: Context,
        info: GraphQLResolveInfo
    ): AsyncIterator<R | Result>;
    resolve?<R = Result, P = Parent>(
        parent: P,
        args: Args,
        context: Context,
        info: GraphQLResolveInfo
    ): R | Result | Promise<R | Result>;
};

export interface Query {
    grid: Grid;
}

export interface Grid {
    id: string;
    gridItems: GridItem[][];
    currentPlayer: Player;
    winner?: Player | null;
    isFinished: boolean;
    size: number;
}

export interface GridItem {
    id: string;
    player?: Player | null;
}

export interface Mutation {
    executeTurn: Grid;
    newGame: Grid;
}
export interface GridQueryArgs {
    id: string;
}
export interface ExecuteTurnMutationArgs {
    id: string;
    player: string;
    x: number;
    y: number;
}

export enum Player {
    NAUGHT = "NAUGHT",
    CROSS = "CROSS"
}

export namespace QueryResolvers {
    export interface Resolvers<Context = any> {
        grid?: GridResolver<Grid, any, Context>;
    }

    export type GridResolver<R = Grid, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context,
        GridArgs
    >;
    export interface GridArgs {
        id: string;
    }
}

export namespace GridResolvers {
    export interface Resolvers<Context = any> {
        id?: IdResolver<string, any, Context>;
        gridItems?: GridItemsResolver<GridItem[][], any, Context>;
        currentPlayer?: CurrentPlayerResolver<Player, any, Context>;
        winner?: WinnerResolver<Player | null, any, Context>;
        isFinished?: IsFinishedResolver<boolean, any, Context>;
        size?: SizeResolver<number, any, Context>;
    }

    export type IdResolver<R = string, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context
    >;
    export type GridItemsResolver<
        R = GridItem[][],
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type CurrentPlayerResolver<
        R = Player,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type WinnerResolver<
        R = Player | null,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type IsFinishedResolver<
        R = boolean,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type SizeResolver<
        R = number,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}

export namespace GridItemResolvers {
    export interface Resolvers<Context = any> {
        id?: IdResolver<string, any, Context>;
        player?: PlayerResolver<Player | null, any, Context>;
    }

    export type IdResolver<R = string, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context
    >;
    export type PlayerResolver<
        R = Player | null,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}

export namespace MutationResolvers {
    export interface Resolvers<Context = any> {
        executeTurn?: ExecuteTurnResolver<Grid, any, Context>;
        newGame?: NewGameResolver<Grid, any, Context>;
    }

    export type ExecuteTurnResolver<
        R = Grid,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context, ExecuteTurnArgs>;
    export interface ExecuteTurnArgs {
        id: string;
        player: string;
        x: number;
        y: number;
    }

    export type NewGameResolver<
        R = Grid,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}
