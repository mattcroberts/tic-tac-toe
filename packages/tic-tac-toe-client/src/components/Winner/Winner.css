.root {
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
}

.inner {
    padding: 50px 80px;
    background-color: #d0d0d0;
    color: #000;
    font-size: 80px;
    font-weight: 800;
}

.naught {
    composes: naught from "../GridItem/GridItem.css";
}

.cross {
    composes: cross from "../GridItem/GridItem.css";
}
