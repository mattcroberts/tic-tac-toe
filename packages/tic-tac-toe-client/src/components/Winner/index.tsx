import * as React from "react";
import classnames from "classnames";

import { Player as IPlayer, Symbol as ISymbol } from "../../../typings/types";
import style from "./Winner.css";
import getPlayerSymbol from "../../utils/playerMap";

const Winner = ({ winner }: { winner?: IPlayer | null }) => {
    const text = winner
        ? [
              <span
                  className={classnames({
                      [style.naught]: winner.symbol === ISymbol.NAUGHT,
                      [style.cross]: winner.symbol === ISymbol.CROSS
                  })}
                  key={"winner"}
              >
                  {getPlayerSymbol(winner.symbol)}
              </span>,
              " has won!"
          ]
        : null;
    return (
        <div className={style.root}>
            <span className={style.inner}>{text || "Draw"}</span>
        </div>
    );
};

export default Winner;
