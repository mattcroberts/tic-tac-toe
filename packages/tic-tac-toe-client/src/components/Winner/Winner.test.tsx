import * as React from "react";
import { shallow } from "enzyme";

import Winner from "./index";
import { Symbol as ISymbol } from "../../../typings/types";

describe("Winner", () => {
    const defaultProps = {};

    it("should render draw", () => {
        const component = shallow(<Winner {...defaultProps} />);

        expect(component).toMatchSnapshot();
    });

    it("should render X win", () => {
        const component = shallow(
            <Winner
                {...defaultProps}
                winner={{ id: "123", symbol: ISymbol.CROSS }}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("should render O win", () => {
        const component = shallow(
            <Winner
                {...defaultProps}
                winner={{ id: "123", symbol: ISymbol.NAUGHT }}
            />
        );

        expect(component).toMatchSnapshot();
    });
});
