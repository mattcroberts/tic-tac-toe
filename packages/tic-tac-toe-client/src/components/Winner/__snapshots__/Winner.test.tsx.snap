// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Winner should render O win 1`] = `
<div
  className="root"
>
  <span
    className="inner"
  >
    <span
      className="naught"
      key="winner"
    >
      o
    </span>
     has won!
  </span>
</div>
`;

exports[`Winner should render X win 1`] = `
<div
  className="root"
>
  <span
    className="inner"
  >
    <span
      className="cross"
      key="winner"
    >
      x
    </span>
     has won!
  </span>
</div>
`;

exports[`Winner should render draw 1`] = `
<div
  className="root"
>
  <span
    className="inner"
  >
    Draw
  </span>
</div>
`;
